Code fragments in DevStudio

Arthoni opened the pull request.

jctophefabre commented on src/openfluid/config.hpp.in:
```diff
@@ -130,6 +130,9 @@ const std::string BUILDEREXTS_WARESBIN_STD_PATH =
 
 
+// Fragments dirs
+const std::string FRAGMENTS_PATH = "@OPENFLUID_FRAGMENTS_PATH@";
```
Variable to rename `WARESDEV_SRC_FRAGMENTSDIR` and move it near line 164

Arthoni replied:
This variable name has a wider use than the name of the subdirectory in the ware source, it is also the location name used in the `waresdev/` directory  (and for workspace) for shared fragments, which also explains why it is line 49 in CMake.in.cmake.

jctophefabre replied:
OK, in this case, the correct name would be `WARESDEV_FRAGMENTS_DIR` as fragments are only related to wares development, and it is just a directory name (not a path). And it should be moved to the "waresdev" part of this file for better grouping and consistency of variables.
(and I am totally aware that a full review of variables naming in this file should be done, as there are
 some consistency errors)

jctophefabre commented on src/openfluid/ui/waresdev/AbstractSrcImportDialog.cpp:
```diff
@@ -0,0 +1,354 @@
+
+AbstractSrcImportDialog::~AbstractSrcImportDialog()
+{
+  delete mp_SrcImportSequenceManager;
```
Deletion may crash if pointer is nullptr. Why not using a smart pointer?

jctophefabre commented on src/openfluid/ui/waresdev/AbstractSrcImportDialog.hpp:
```diff
@@ -0,0 +1,143 @@
+  protected:
+
+    SrcImportSequenceManager* mp_SrcImportSequenceManager = nullptr;
+    openfluid::ui::waresdev::HubManager* mp_HubManager = nullptr;
```
Why not using a smart pointer?

jctophefabre commented on src/openfluid/ui/waresdev/GitImportWorker.hpp:
```diff
@@ -0,0 +1,117 @@
+
+    bool m_SslNoVerify;
+
+
```
Too much empty lines (should be one only in this case)

jctophefabre commented on src/openfluid/ui/waresdev/GitImportWorker.hpp:
```diff
@@ -0,0 +1,117 @@
+
+
+    std::vector<std::pair<QString, QString>> m_ElementsToImport; // first: Git URL, second: local path
+
```
Two empty lines here 😉

jctophefabre commented on src/openfluid/utilsq/GitProxy.cpp:
```diff
@@ -178,16 +175,28 @@ void GitProxy::processErrorOutputAsInfo()
   }
 
+  // Manual check if .git/index.lock detected (ongoing process)
+  if (!WorkingDirectory.isEmpty())
```
Mark this with HACK or TOIMPL tag

jctophefabre commented on src/openfluid/utilsq/RESTClient.cpp:
```diff
@@ -155,13 +155,15 @@ RESTClient::RESTClient()
 
 void RESTClient::removeRawHeader(const QByteArray Key)
 {
+  t_HeadersList NewHeaders;
```
Why these modifications?

Arthoni replied:
I assume it was to avoid changing the list i am iterating on, but not sure anymore if this was really necessary.

jctophefabre commented on src/openfluid/waresdev/WareSrcHelpers.hpp:
```diff
@@ -0,0 +1,80 @@
+  @param[in] WarePath the ware path
+  @return true if ware found in workspace
+*/
+inline bool isWareInCurrentWorkspace(const std::string& WarePath)
```
Why this helper function? Is it a simple shortcut?

Arthoni replied:
yes, considering the same operation is done twice with access to a member of an object from an other function, the use of a shortcut seemed relevant

Arthoni commented:
# Details on the 2 post-review commits 
(following the order of files in the commit diff of GitHub, commit by commit)

##  Structural improvements for fragments management

- Expanded tests
	- resources/tests/api/MultiAPI_TEST_data.json : adding cases about fragments to test API output

- Adjustments related with PR feedback by JC
	- src/openfluid/config.hpp.in: changed cmake var for fragment dir
	- src/openfluid/ui/waresdev/AbstractSrcImportDialog.cpp: replaced m_HubManager from raw pointer to direct object instance
	- src/openfluid/ui/waresdev/HubManager.cpp
		- hub manager is persistent but url injection reset all values
		- changed `isConnected` check to new logic
		- two functions for connection to hub: connectToHub / asyncConnectToHub, for synchronous and asynchronous (threaded worker) cases
		- added constructor for hub manager with empty hub url
		- connectToHub: calls internal worker class to get a temporary instance (with "this" as parent value) and run it synchronously, returns worker status as bool, was called "connect" before but changed to avoid ambiguity with Qt connect function
		(connect/disconnect were not changed in FluidHubAPIClient since it is not a Qt class)
		- asyncConnectToHub: creates an internal worker pointer to send it to the thread, binds relevant signals for start / end / delete, and starts the worker
		
		- definition of `run`, the only function of the HubManager::HubConnectWorker: get context information from HubManager instance saved in "parent" attribute, and use the HubClient of the parent to connect to URL and save found wares and fragments in parent containers. emits `finished` with bool for success and string for status information (used in asynchronous case), and returns as bool the success status (used in synchronous case)
	- src/openfluid/ui/waresdev/HubManager.hpp
		- Worker is defined as private internal class since the only one to use it is the HubManager
		- Split manager/worker is necessary for thread management and bool return logic

- Updates tests (even commented parts)


##  Gitaskpass handling for username injection 
- CMake.in.cmake
	- Added env var `OFBUILD_GITASKUSER_ENVVAR_PREFIX` to contain username variable name, built with same suffix than askpass_var
- src/apps/openfluid-gitaskpass/main.cpp
	- caught incoming arguments to detect what to display, send password from env in general case, but send username from env if first arg contains "Username". 
	
:warning: Identified problem: username injection will break if git displays message in an other language than english.

- src/openfluid/utilsq/GitProxy.cpp
	- added parameter `UsernameViaEnv` to `launchCommand` to chose between username in URL and username in env, default is `false` <=> username in URL
	- only use case for username in env is submodule call (other cases like `git clone` keep previous username in URL)

jctophefabre commented on resources/tests/api/MultiAPI_TEST_data.json:
```diff
@@ -319,7 +322,7 @@
               {
                 "description": "This is sim.staging.example2...", 
-                "git_url": "http://147.100.175.211:8181/git-service/wares/simulators/sim.staging.example2", "users_ro": ["jbuckley", "kcobain", "dgrohl"], "users_rw": ["dgrohl"], "ware_url": "http://147.100.175.211:8181/api/wares/simulators/sim.staging.example2", "git_branches": [], 
+                "git_url": "http://147.100.175.211:8181/git-service/wares/simulators/sim.staging.example2", "users_ro": ["jbuckley", "kcobain", "dgrohl"], "users_rw": ["dgrohl"], "ware_url": "http://147.100.175.211:8181/api/wares/simulators/sim.staging.example2",
```
Do not use real URL (potentially reveals internal secrets 😉)

jctophefabre commented on CMake.in.cmake:
```diff
@@ -44,6 +44,7 @@ SET(OPENFLUID_DEFAULT_OUTPUT_PATH "OPENFLUID.OUT")
 SET(OPENFLUID_WORKSPACE_PATH "workspace")
 SET(OPENFLUID_WARESBIN_PATH "wares")
 SET(OPENFLUID_WARESDEV_PATH "wares-dev")
+SET(OPENFLUID_SHARED_PATH "shared")
```
Confusing name
